genesyscloud/architect_user_prompt/genesyscloud_architect_user_prompt_proxy.go

```diff
@@ -0,0 +1,199 @@
+	response, err := p.architectApi.DeleteArchitectPrompt(id, allResources)
+	if err != nil {
+		return response, fmt.Errorf("failed to delete architect user prompt: %s", err)
+	}
+	return response, nil
+}
+
+func getAllArchitectUserPromptsFn(ctx context.Context, p *architectUserPromptProxy, includeMediaUris bool, includeResources bool, nameArr []string) (*[]platformclientv2.Prompt, *platformclientv2.APIResponse, error, bool) {
+	var (
+		pageCount  int
+		pageNum    = 1
+		allPrompts []platformclientv2.Prompt
+	)
+
+	const pageSize = 100
+	userPrompts, response, err := p.architectApi.GetArchitectPrompts(pageNum, pageSize, nameArr, "", "", "", "", includeMediaUris, includeResources, nameArr)
+
+	if err != nil {
+		return nil, response, fmt.Errorf("failed to get list of prompts: %v", err), true
+	}
+
+	if userPrompts != nil && userPrompts.Entities != nil && len(*userPrompts.Entities) > 0 {
+		allPrompts = append(allPrompts, *userPrompts.Entities...)
+	}
+
+	pageCount = *userPrompts.PageCount
+
+	for pageNum := 2; pageNum <= pageCount; pageNum++ {
+		userPrompts, response, getErr := p.architectApi.GetArchitectPrompts(pageNum, pageSize, nameArr, "", "", "", "", includeMediaUris, includeResources, nameArr)
+		if getErr != nil {
+			return nil, response, fmt.Errorf("failed to get page of prompts: %v", getErr), true
+		}
+
+		if userPrompts.Entities == nil || len(*userPrompts.Entities) == 0 {
+			break
+		}
+	}
+
+	return &allPrompts, response, nil, false
+}
+
+func createArchitectUserPromptResourceFn(ctx context.Context, p *architectUserPromptProxy, id string, body platformclientv2.Promptassetcreate) (*platformclientv2.Promptasset, *platformclientv2.APIResponse, error) {
+	promptAsset, response, err := p.architectApi.PostArchitectPromptResources(id, body)
+	if err != nil {
+		return nil, response, fmt.Errorf("failed to create architect user prompt resource: %s", err)
+	}
+	return promptAsset, response, nil
+}
+
+func updateArchitectUserPromptResourceFn(ctx context.Context, p *architectUserPromptProxy, id string, languageCode string, body platformclientv2.Promptasset) (*platformclientv2.Promptasset, *platformclientv2.APIResponse, error) {
+	promptAsset, response, err := p.architectApi.PutArchitectPromptResource(id, languageCode, body)
+	if err != nil {
+		return nil, response, fmt.Errorf("failed to update architect user prompt resource: %s", err)
+	}
+	return promptAsset, response, nil
+}
+
+func getArchitectUserPromptIdByNameFn(ctx context.Context, p *architectUserPromptProxy, nameArr []string) (string, *platformclientv2.APIResponse, error, bool) {
+	// Query user prompt by name. Retry in case search has not yet indexed the user prompt.
+	prompts, response, err, retryable := p.getAllArchitectUserPrompts(ctx, true, true, nameArr)
+	if err != nil {
+		return "", response, fmt.Errorf("error requesting user prompt by name %s: %s", nameArr, err), retryable
+	}
+	if prompts == nil || len(*prompts) == 0 {
```

Comment:
You don't need to check if `prompts == nil`, `getAllArchitectUserPrompts` always returns a value as long as there's no error. You don't need to check if the length is 0 either, if it is the loop just won't run and the error will be returned

Reply:
"always returns a value" that is a very naive considering publicAPI isn't governed by us nor they have always stuck to API contracts.

Reply:
That would be true if you were calling the api here but you're calling `getAllArchitectUserPrompts`. `getAllArchitectUserPrompts` always returns a pointer to a slice even if the slice it points to is nil as long, as there's no error, meaning this nil check will never be true. In a situation where the under lying slice is nil it is handled by the range function and the loop won't run

Reply:
i was referring to the speculation about what is "always going to be return" slice etc... as we don't own the service, hence its better to handle edge scenarios.

Reply:
Even if the service doesn't return any data, `prompts` still can't be nil. The way `getAllArchitectUserPrompts` is written, as long as there is no error it will always return a value for `prompts` since  `prompts` is a pointer to an array even if the array is nil we can still create a pointer to it. So basically just because an array is nil, doesn't mean a pointer to that array is

Reply:
if someone does change what the getAllArchitectUserPrompts returns